keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1578 @@
+Secret, EmptyDir, etc), CSI plugins, container runtime (which in turn invokes
+CNI plugins) and associated runtime handler and other components. This KEP
+proposes a `SandboxReady` condition in pod status to indicate successful
+completion of pod sandbox creation by Kubelet. The `SandboxReady` condition will
+mark an important milestone in the pod's lifecycle similar to `ContainersReady`
+and the overall `Ready` conditions in pod status today.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today, the scheduler surfaces a specific pod condition: `PodScheduled` that
+clearly identifies whether a pod got scheduled by the scheduler and when
+scheduling completed. However, no specific conditions around initialization of
+successfully scheduled pods from the perspective of completion of pod sandbox
+creation is surfaced to cluster administrators in a scoped and consumable
+fashion.
+
+There is an existing pod condition: `Initialized` that tracks execution of init
+containers. For pods without init containers, the `Initialized` condition is set
+when the Kubelet starts to process a pod before any sandbox creation activities
+start. For pods with init containers, the `Initialized` condition is set when
+init containers have been pulled and executed to completion. Therefore, the
+existing `Initialized` condition is insufficient and inaccurate for tracking
+completion of sandbox creation of all pods in a cluster. This distinction
+becomes especially relevant in multi-tenant clusters where individual tenants
+own the pod specs (including the set of init containers) while the cluster
+administrators are in charge of storage plugins, networking plugins and
+container runtime handlers.
+
+A dedicated condition around readiness of the pod sandbox will benefit cluster
+operators (especially of multi-tenant clusters) who are responsible for
+configuration and operational aspects of the various components that play a role
+in pod sandbox creation: CSI plugins, CRI runtime and associated runtime
+handlers, CNI plugins, etc. The duration between `lastTransitionTime` field of
+the `SandboxReady` condition (with `status` set to `true` for a pod for the
+first time) and the existing `PodScheduled` condition will allow metrics
+collection services to surface total latency of all the components involved in
+pod sandbox creation as an SLI. Cluster operators can use this to publish SLOs
+around pod initialization to their customers who launch workloads on the
+cluster.
+
+Custom pod controllers/operators can use a dedicated condition indicating
```

Comment:
I think the only thing you can pro-actively do is delete a pod that is not having its Sandbox go ready, but the underlying reason for why it did not go ready may vary by reason (secret does not exist, configmap does not exist, pull secret is invalid, volumes cannot be attached).  Let me know if I am misunderstanding.

Reply:
Deleting the pod and re-creating is absolutely necessary if a pod is not coming up - there is no avoiding that. What I was trying to allude to here is: for controllers whose custom resources specify VolumeClaimTemplates, the extra condition can serve as a signal on whether to try to also delete existing PVCs (that the failing pod mounts) and re-create them versus just re-creating the pod that is failing to come up. If the new condition is reporting `true` for failed pods, pod re-creation is all that needs to be attempted. If the new condition is reporting `false`, there is most likely a deeper underlying problem - for example, PVs (bound to the PVCs spun up based on templates) are not attaching/mounting and the controller may want to try to recreate those as well to get the pod to successfully come up.